### DISCUSSION TITLE
CompleteHamiltonian2

### DISCUSSION THREAD

**Discussion by Jamiewhbee** [Part-II-Computational-Physics/.github#27] [Q&A] [answered]
I want to evaluate this in a similar way to CompleteHamiltonian. I think this would involve defining a set of N-2 copies of the 2x2 identity matrix I, and then adding the 4x4 LocalHamiltonian matrix to the set (rather than replacing two of the N copies of the identity matrix, as is done in CompleteHamiltonian). However, this involves somehow combining matrices with different shapes, and I'm not sure how to do this. I'm also not sure if this is even the right approach.

**Comment by Jamiewhbee** [comment, marked as answer]
I've given up on this approach and currently I'm trying to evaluate H by summing np.kron(np.kron(np.identity(2**i),LocalHamiltonian(h)),np.identity(2**(N-i-2))) from i = 0 to i = N-2, but I think I need an N-1th term in the sum to account for periodic boundary conditions, and I don't know how to implement this.

**Reply by Jamiewhbee** [reply]
Formatting messed that up, the code looks like this
![image](https://user-images.githubusercontent.com/122116209/221286517-a8ab83ab-bbf5-44dc-a50c-ba65933e01d8.png)

**Reply by Jamiewhbee** [reply]
The matrices produced look like this (top = CompleteHamiltonian, bottom = my current CompleteHamiltonian2)

**Reply by Jamiewhbee** [reply]
![image](https://user-images.githubusercontent.com/122116209/221287261-7d8a0d0f-4b4e-4681-951f-3cff1a3929fb.png)

**Reply by Jamiewhbee** [reply]
I've got it working by adding this line, but I don't know why the factor is N-1 specifically
![image](https://user-images.githubusercontent.com/122116209/221297223-1a343956-6841-471f-a3aa-927c56732c43.png)

**Comment by JNHallen** [comment]
Hi! The best approach for this task is indeed to construct identity matrices of size 2^k, and 2^(N-k-2) and then combining these with a Kronecker product like you are doing. This is just constructing the Hamiltonian as a sum of the local Hamiltonian. 
$$ {\cal H}_{\mathrm{Ising}} = \sum_{i=1}^N {\cal H}_{i, i+1}^{\mathrm{loc}} $$

For the final term here, ${\cal H}_{N, 1}^{\mathrm{loc}} $, the easiest thing is actually to not use the LocalHamiltonian matrix you have constructed, but to instead just apply the necessary Pauli matrices making up this term. That is what your final line is doing, although I would maybe suggest splitting it up over a couple of lines to make things clearer. Of course, if you want to apply a single Pauli matrix at site 1 to a system of $N$ spins, that is actually going to be the Kronecker product of this matrix with $N-1$ $(2\times 2)$ identity matrices. That is where the factor of $N-1$ is coming from.
